api/graphql/reservations/reservation_serializers.py

```diff
@@ -717,34 +721,142 @@ def check_metadata_fields(self, data, reservation_unit) -> None:
+        else:
+            return PaymentType.ONLINE
+
+    def validate(self, data):
+        data = super().validate(data)
+
+        if self.instance.payment_order.exists():
+            raise ValidationErrorWithCode(
+                "Reservation cannot be changed anymore because it is attached to a payment order",
+                ValidationErrorCodes.CHANGES_NOT_ALLOWED,
+            )
+
+        if self.instance.reservation_unit.count() > 1:
+            raise ValidationErrorWithCode(
+                "Reservations with multiple reservation units are not supported.",
+                ValidationErrorCodes.MULTIPLE_RESERVATION_UNITS,
+            )
+
+        if not self._requires_handling():
+            payment_type = data.get("payment_type", "").upper()
+            reservation_unit = self.instance.reservation_unit.first()
+
+            if not reservation_unit.payment_product:
+                raise ValidationErrorWithCode(
+                    "Reservation unit is missing payment product",
+                    ValidationErrorCodes.MISSING_PAYMENT_PRODUCT,
+                )
+
+            if not payment_type:
+                data["payment_type"] = self._get_default_payment_type()
+            elif not reservation_unit.payment_types.filter(code=payment_type).exists():
+                allowed_values = list(
+                    map(lambda x: x.code, reservation_unit.payment_types.all())
+                )
+                raise ValidationErrorWithCode(
+                    f"Reservation unit does not support {payment_type} payment type. "
+                    f"Allowed values: {', '.join(allowed_values)}",
+                    ValidationErrorCodes.INVALID_PAYMENT_TYPE,
+                )
+        return data
 
     @property
     def validated_data(self):
         validated_data = super().validated_data
 
-        reservation_unit_requires_handling = self.instance.reservation_unit.filter(
-            require_reservation_handling=True
-        ).exists()
-
-        if (
-            reservation_unit_requires_handling
-            or self.instance.applying_for_free_of_charge
-        ):
+        if self._requires_handling():
             validated_data["state"] = STATE_CHOICES.REQUIRES_HANDLING
         else:
             validated_data["state"] = STATE_CHOICES.CONFIRMED
+
         return validated_data
 
     def save(self, **kwargs):
+        self.fields.pop("payment_type")
         instance = super().save(**kwargs)
 
+        state = self.validated_data["state"]
+        if state == STATE_CHOICES.CONFIRMED:
+            payment_type = self.validated_data["payment_type"].upper()
+            price_net = round_decimal(self.instance.price_net, 2)
+            price_vat = round_decimal(
+                self.instance.price_net * (self.instance.tax_percentage_value / 100), 2
+            )
+            price_total = round_decimal(self.instance.price, 2)
+
+            if payment_type == PaymentType.ON_SITE:
```

Comment:
These order creations could be inside the payments app, could they? 🤔

Reply:
I started to move them there but I when I was almost done, I got a feeling that the code felt really out-of-place. 

Creating those orders is kinda specific to the confirm step and having this logic to be in the `merchants` app didn't feel ...right. If someone later opens those helper functions or whatnot, the code looks really confusing because it is out of the context.

So I decided to keep them here, but if you feel that this really isn't the place, I can move them ☺️